byron/ledger/executable-spec/test/Ledger/HasTypeReps/Properties.hs

```diff
@@ -0,0 +1,170 @@
+import           Test.Tasty.HUnit                   (Assertion, testCase, (@?=))
+
+--------------------------------------------------------------------------------
+-- Example HasTypeReps.typeReps for TxIn, Tx
+--------------------------------------------------------------------------------
+
+aTx :: Tx
+aTx = undefined
+
+aTxId :: TxId
+aTxId = TxId (hash aTx)
+
+-- | 'TxIn' has a generic instance for 'HasTypeReps', which recursively adds
+--   'typeReps' for all types within 'TxIn'.
+exampleTypeRepsTxIn :: Assertion
+exampleTypeRepsTxIn =
+  let txIn = TxIn aTxId 0
+  in (typeReps txIn) @?= typeOf (undefined::TxIn)
+                         <| typeOf (undefined::TxId)
+                         <| typeOf (undefined::Hash)
+                         <| typeOf (undefined::Int)
+                         <| typeOf (undefined::Natural)
+                         <| empty
+
+-- | A 'TxWits' term may contain multiple inputs/outputs/witnesses.
+--   In this example, we have 2 inputs and show how the 'typeReps' for
+--   'TxIn' is repeated twice.
+exampleTypeRepsTx :: Assertion
+exampleTypeRepsTx =
+  let
+    (in0,in1) = (TxIn aTxId 0, TxIn aTxId 1)
+    outs = []
+    wits = []
+    tx = TxWits (Tx [in0, in1] outs) wits
+  in (typeReps tx) @?= typeOf (undefined::TxWits)
+                       <| typeOf (undefined::Tx)
+                       <| typeOf (undefined::[TxIn])
+                       <| typeReps in0
+                       >< typeReps in1
+                       >< (Seq.fromList [ typeOf (undefined::[TxOut])
+                                        , typeOf (undefined::[Wit]) ])
+
+--------------------------------------------------------------------------------
+-- Properties of abstractSize of TxWits / TxIn /TxOut / Wit
+--------------------------------------------------------------------------------
+
+-- | Since 'VKey' appears in both Wit and TxOut (as part of Addr), we
+--   need to use the same "cost" for this type in the following tests.
+vKeyCost :: Size
+vKeyCost = 17
+
+-- | Example costs for types in a 'TxIn'.
+costsTxIn :: Map TypeRep Size
+costsTxIn = Map.fromList [ (typeOf (undefined :: TxIn), 2)
+                         , (typeOf (undefined :: TxId), 3)
+                         , (typeOf (undefined :: Hash), 5) ]
+
+-- | Example costs for types in a 'TxOut'.
+costsTxOut :: Map TypeRep Size
+costsTxOut = Map.fromList [ (typeOf (undefined :: TxOut), 11)
+                          , (typeOf (undefined :: Addr), 13)
+                          , (typeOf (undefined :: VKey), vKeyCost)
+                          , (typeOf (undefined :: Lovelace), 23) ]
+
+-- | Example costs for types in a 'Wit'.
+costsWit :: Map TypeRep Size
+costsWit = Map.fromList [ (typeOf (undefined :: Wit), 31)
+                        , (typeOf (undefined :: VKey), vKeyCost)
+                        , (typeOf (undefined :: Sig Tx), 41)]
+
+-- | This property tests that 'abstractSize' accounts for multiple occurences
```

Comment:
Now I see why you needed to choose different numbers. I like the idea behind this property.

Reply:
On a second thought, I don't think it is in this property where you need the different abstract sizes :smiley:  I'll read on ...